teachertool/src/services/makecodeEditorService.ts

```diff
@@ -0,0 +1,52 @@
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+let makecodeEditorRef: HTMLIFrameElement | undefined;
+const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+let nextId: number = 0;
+let pendingMessages: {[index: string]: PendingMessage} = {};
+
+const onMessageReceived = (event: MessageEvent) => {
+    const data = event.data as pxt.editor.EditorMessageRequest;
+    if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+        const pending = pendingMessages[data.id];
+        pending.handler(data);
+        delete pendingMessages[data.id];
+        return;
+    }
+}
+
+export const sendMessageAsync = (message?: any) => {
+    return new Promise(resolve => {
+        const sendMessageCore = (message: any) => {
+            message.response = true;
+            message.id = nextId++ + "";
+            pendingMessages[message.id] = {
+                original: message,
+                handler: resolve
+            };
+            makecodeEditorRef!.contentWindow!.postMessage(message, "*");
+        }
+
+        if (makecodeEditorRef) {
```

Comment:
I'm not sure this check is sufficient to ensure the message doesn't get sent to the editor before it is ready.

Reply:
Agree. We'll probably need to listen for some "loaded" event (not sure what that is yet) and then set a flag, rather than checking the ref directly.

Reply:
For a more "sufficient" check, we may need the event I'm working on adding to detect when the editor is done loading. If that's the case, I wouldn't block on this. I'm building off of your changes, so I can add it once you check in if you're otherwise ready to go. If you're still working on other things though, I can try to separate it out and send it in a new PR. (I probably should've done that to begin with, sorry...)

Reply:
Yeah, I was mulling on this and waiting to respond because I kinda figured that was the case. I also think that the if statement is a pretty weak protection, but I don't know if there's anything right now that I can do to make this better.

Maybe we could also have some sort of async function to wait for the editor to load? But that sounds a bit messy to me.